### PR TITLE
Feature/DH-599 update company interactions

### PR DESCRIPTION
**Pull request by ztolley**
screenshot from 2017-09-11 18-28-45

* Updates company to use collections to list related interactions.
* Switches interaction data to be collected from new V3 endpoint
* Introduce the start of moving company to middleware pattern by using middleware to get company details, labels and interaction data before render.
* Remove code no longer needed in getExpandedDitCompany

### CODE REVIEW

**Comment by teneightfive** [comment]
Once thing I just noticed on the screenshot - should we move the _Add interaction_ button into the same location it is in for investment projects:

image

File: src/apps/companies/middleware/details.js [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+  if (company.companies_house_data && company.companies_house_data !== null) {
+    return getFormattedAddress(company.companies_house_data, 'registered')
+  }
+
+  return getFormattedAddress(company, 'registered')
+}
+
+function getHeadingName (company) {
+  if (company.trading_name && company.trading_name.length > 0) {
```
Comment (teneightfive) [inline src/apps/companies/middleware/details.js]: This feels like something that could just be handled in the view as the only place it is used is when being passed to `res.locals`.

File: src/apps/companies/middleware/details.js [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+      next(error)
+    }
+  }
+
+  next()
+}
+
+function getHeadingAddress (company) {
+  // If this is a CDMS company
+  const cdmsTradingAddress = getFormattedAddress(company, 'trading')
```
Comment (teneightfive) [inline src/apps/companies/middleware/details.js]: I'd be tempted to also consider handling this in the view as its only for being used for rendering.

It then also allows us to use the view filters rather than having separate library methods for it.

File: src/apps/companies/views/interactions.njk [hunk 3]
```diff
@@ -1,84 +1,25 @@
...
-  {% elseif (not company.contacts or company.contacts.length == 0) %}
+  <div class="split-header">
     <h2 class="heading-medium">Company interactions</h2>
-    <p>
-      An interaction could be a meeting, call, email or another activity where you have been in touch with a company.
-    </p>
-    <p id="no-contact-warning" class="infostrip">
-      You currently have no contacts for this company. To add an interaction you must first <a href="{{addContact}}">add a contact</a>.
+    {% if (company.contacts and company.contacts.length > 0) %}
+      <a id="add-interaction-link" class="button" href="{{addInteractionUrl}}">Add interaction</a>
```
Comment (teneightfive) [inline src/apps/companies/views/interactions.njk]: Maybe move this to the same location as on investment projects.

File: src/apps/interactions/transformers.js [hunk 4]
```diff
@@ -40,6 +40,42 @@ function transformInteractionToListItem ({
   }
 }
 
+function transformInteractionToCompanyListItem ({
```
Comment (teneightfive) [inline src/apps/interactions/transformers.js]: This looks like quite a repetition of the other transformer in this file. Is the only different to remove company from the meta?

Maybe consider chaining another transformer to remove that meta item rather than duplicating it.

If the transformer in search/collection changed you'd have to make the change twice or risk it being missed in this listing.

File: src/apps/companies/middleware/details.js (hunk 2 above)
Comment (ztolley) [reply src/apps/companies/middleware/details.js]: Good point though not trivial

File: src/apps/companies/middleware/details.js (hunk 1 above)
Comment (ztolley) [reply src/apps/companies/middleware/details.js]: Easier to test in the middleware

File: src/apps/interactions/transformers.js (hunk 4 above)
Comment (ztolley) [reply src/apps/interactions/transformers.js]: Yeah I kept thinking that, bugged me. will do

File: src/apps/companies/middleware/details.js (hunk 2 above)
Comment (ztolley) [reply src/apps/companies/middleware/details.js]: Can we agree that I'll create a filter for it later and refactor this to use it?

File: src/apps/companies/middleware/details.js (hunk 1 above)
Comment (teneightfive) [reply src/apps/companies/middleware/details.js]: I don't think we need to unit test views.

What is rendered is something that can be picked up by end to end tests.

This feels like the wrong place to have this logic though.

Comment (ztolley) [reply src/apps/companies/middleware/details.js]: Well we could argue that the view should not contain logic, but simply show values and loop.

I was simply trying to start to introduce middleware into company as part of chipping away at the current way it's structured. If I move this into the view layer then it's more risky and I then have to go back and test it with companies house vs cdms records

Can we discuss this with the other Fed's and if we want to do this, and the address suggestion, we can do it as part of a separate PR that would focus on that.

File: src/apps/companies/middleware/details.js (hunk 2 above)
Comment (teneightfive) [reply src/apps/companies/middleware/details.js]: Yes, want to create an issue so we don't forget?

File: src/apps/interactions/transformers.js [hunk 5]
```diff
@@ -40,6 +40,24 @@ function transformInteractionToListItem ({
...
+ *
+ * @param {object} iteraction action (id, type, name, meta)
+ *
+ * Accepts an iteraction collection obect and reduces it to exclude company data
+ */
+function transformInteractionToCompanyListItem (interaction) {
+  const { id, type, name, meta } = transformInteractionToListItem(interaction)
+
+  return Object.assign({}, {
+    id,
```
Comment (teneightfive) [inline src/apps/interactions/transformers.js]: I'm wondering if this is the best implementation as it feels like it's still repeating some things.

What about chaining the `.then()` in the company controller to strip out the unwanted value.

Whilst writing this and looking at investment project listing...why don't we just leave this in? We have done for investment projects. That way the other meta items will also still appear in the same location therefore making them easier to scan.

File: test/unit/apps/companies/controllers/interactions.test.js [hunk 6]
```diff
@@ -1,232 +1,65 @@
-const { render } = require('~/test/unit/nunjucks')
-const next = function (error) {
-  throw Error(error)
-}
-
 describe('Company interactions controller', function () {
-  let company
-  let companyinteractioncontroller
+  beforeEach(() => {
+    this.sandbox = sinon.sandbox.create()
```
Comment (teneightfive) [inline test/unit/apps/companies/controllers/interactions.test.js]: Have been wondering whether to move this out to a global level so that the sandbox is available in all tests. What are the thoughts on that?

Definitely not part of this change but seeing the diff reminded me. @tyom and I were discussing that we should define the pattern for our tests too as there's a lot of different methods going on at the moment.

File: src/apps/interactions/transformers.js (hunk 5 above)
Comment (ztolley) [reply src/apps/interactions/transformers.js]: The stripping should happen in the transformer.

The reason for having it, and in the PR for contacts, is that they would all just say this company. 

I tried chaining them but that didn't work plus the controller or middleware needs to know to do that, so poor abstraction.

File: src/apps/companies/views/interactions.njk [hunk 7]
```diff
@@ -1,84 +1,25 @@
 {% extends "./_layout-view.njk" %}
 
+{% set addButton %}
+  <a id="add-interaction-link" class="button" href="{{addInteractionUrl}}">Add interaction</a>
```
[the remark this answers is not in the document]
Comment (ztolley) [reply src/apps/companies/views/interactions.njk]: Also updated investment interactions to be consistent.

Also no longer generate the url in the controller, just create it in the view.

https://github.com/uktrade/data-hub-frontend/pull/571

[the remark this answers is not in the document]
Comment (ztolley) [reply src/apps/companies/views/interactions.njk]: The id is used for tests

File: test/unit/apps/companies/controllers/interactions.test.js (hunk 6 above)
Comment (ztolley) [reply test/unit/apps/companies/controllers/interactions.test.js]: yes and yes

Tests need patterns too (some use nock, some stub repos)

File: src/apps/interactions/transformers.js (hunk 5 above)
Comment (ztolley) [reply src/apps/interactions/transformers.js]: That all said, ask the others, put it to the vote. it's a quick change

**Comment by ztolley** [comment]
Made a few changes tonight to make names more obvious and improve consistency between this, contact interactions and investment interactions.

File: src/apps/interactions/transformers.js (hunk 5 above)
Comment (teneightfive) [reply src/apps/interactions/transformers.js]: >The reason for having it, and in the PR for contacts, is that they would all just say this company.

Yeah it's not ideal but I noticed whilst writing that comment that investment projects does the same, it still includes the company. I'd also be tempted to say that for now it would also keep the other meta items in the same location as when they appear on search and collections therefore benefitting scanability. 

Was more just a thought to do less work for now and save us some time. Maybe put an issue in GitHub to see if we want to improve it later on, but if we do, we should do the same in investment and probably.

Although looking at contacts when in a company it does also filter out the company name but uses a [different method](https://github.com/uktrade/data-hub-frontend/blob/5b85355be71a47d7f7ab7df0f80974f599a9c391/src/apps/companies/controllers/contacts.js#L18-L23) to reject that meta item. I would say go with this method now for consistency.

File: src/apps/companies/views/interactions.njk (hunk 7 above)
[the remark this answers is not in the document]
Comment (teneightfive) [reply src/apps/companies/views/interactions.njk]: >Also updated investment interactions to be consistent.

I think this is wrong way to update. We've currently used secondary buttons for adding a contact in all of the collections so I would update this PR for consistency rather than investment projects.

## Companies collection

image

## Company search results

image

## Company contacts listing

image

## Company investment project listing

image

[the remark this answers is not in the document]
Comment (teneightfive) [reply src/apps/companies/views/interactions.njk]: >The id is used for tests

@feedmypixel and I were discussing the use of IDs purely for tests yesterday and I feel it's a bad habit to get into. If we need specific hooks we should use prefixed classes (`qa-`) rather than IDs.

IDs should be kept available in case we wish to use them for anchors or for interacting with the ARIA API. Using prefixed classes also better communicates what is going on in the markup and if we wanted we could optionally only show them in certain environments that we run tests on so that we don't pollute the final markup more than we need to.

File: src/apps/interactions/transformers.js (hunk 5 above)
Comment (ztolley) [reply src/apps/interactions/transformers.js]: Thats the old way of doing collections. We should migrate the company contacts to this method instead, not the other way around

Comment (ztolley) [reply src/apps/interactions/transformers.js]: Changing to use the common transformer just to make it simpler and get this done but contact collections in company needs to be re-done

File: src/apps/companies/views/interactions.njk [hunk 8]
```diff
@@ -1,84 +1,25 @@
...
-    </ul>
-  {% elseif (not company.contacts or company.contacts.length == 0) %}
-    <h2 class="heading-medium">Company interactions</h2>
-    <p>
-      An interaction could be a meeting, call, email or another activity where you have been in touch with a company.
-    </p>
-    <p id="no-contact-warning" class="infostrip">
-      You currently have no contacts for this company. To add an interaction you must first <a href="{{addContact}}">add a contact</a>.
-    </p>
+  {% if (company.contacts and company.contacts.length > 0) %}
```
Comment (teneightfive) [inline src/apps/companies/views/interactions.njk]: Checking for just `company.contacts.length` should be enough:

```nunjucks
{% if company.contacts.length %}
  ...
{% endif %}
```

Comment (ztolley) [reply src/apps/companies/views/interactions.njk]: it was being defensive in case there is no property there, and if there isn't it breaks things. tried it before and it did. That said we should be able to trust the api to return contacts no matter what.

Comment (teneightfive) [reply src/apps/companies/views/interactions.njk]: Don't think you need to be in nunjucks. It won't error even if the properties don't exist. It will just fail the condition.

Just tried in a view with a variable and property that didn't exist and it was fine.

Comment (ztolley) [reply src/apps/companies/views/interactions.njk]: done

File: src/apps/companies/middleware/interactions.js [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+    const page = req.query.page || '1'
+    const companyId = req.params.companyId
+
+    res.locals.interactions = await getInteractionsForCompany(token, companyId, page)
+      .then(transformApiResponseToCollection(
+        { entityType: 'interaction' },
+        transformInteractionToListItem
+      ))
+  } catch (error) {
+    next(error)
```
Comment (teneightfive) [inline src/apps/companies/middleware/interactions.js]: Does this need to be returned so that it doesn't call the `next()` without an error on line 20?

Or it might be better to move the `next()` from line 20 up to line 16 within the `try` block.

File: src/apps/interactions/repos.js [hunk 10]
```diff
@@ -49,19 +49,13 @@ function getInteractionsForContact (token, contactId) {
...
-    qs: {
-      company_id: companyId,
-      limit,
-    },
-  })
-  return response.results
+function getInteractionsForCompany (token, companyId, page = 1) {
+  const limit = 10
+  const offset = limit * (page - 1)
+  return authorisedRequest(token, `${config.apiRoot}/v3/interaction?company_id=${companyId}&limit=${limit}&offset=${offset}`)
```
Comment (teneightfive) [inline src/apps/interactions/repos.js]: Just looking at the diff here, why did you change it from passing query strings as an object using a template literal?

Looks easier to read as a querystring object.

Comment (ztolley) [reply src/apps/interactions/repos.js]: All the other methods in the module use this format so I was being consistent

File: src/apps/companies/middleware/interactions.js (hunk 9 above)
Comment (ztolley) [reply src/apps/companies/middleware/interactions.js]: moving it up would mean that a failure further down the chain would come back to this catch, so I'll return the error

Comment (ztolley) [reply src/apps/companies/middleware/interactions.js]: btw we have this in other places.
